### ISSUE TITLE
AWSAPIGateWay - Missing authentication issue. 

### ISSUE THREAD

**Issue by dsonara** [aws-amplify/aws-sdk-ios#1061]
**Describe the bug**
App is not getting successfull api call after hours of in-activity.

**To Reproduce**
1. Creation of Amazon Cognito ID. 
2. API calls.
3. Put app in background.
4. Refresh code after 8 hours. (initializeAmazonCongnitoProviderWithCompletionBlock mentioned below is called first when app comes to foreground only after 8 hours)
5. Missing auth token will appear as suggested in screenshot.

**Which AWS service(s) are affected?**

AWSAPIGateWay

**Expected behavior**

App should not throw AWSPIGateWay error.

**Screenshots**
![error1](https://user-images.githubusercontent.com/11832292/46865131-090be680-ce3a-11e8-9a65-b0f3faa65698.jpg)

**Environment(please complete the following information):**
 - SDK Version: 2.6.2
 - Dependency Manager: Cocoapods
 - Objective C

**Device Information (please complete the following information):**
 - Device: Any iOS device that supports iOS 9+
 - iOS Version: iOS9+
 - Specific to simulators:

**Additional context**
Creation of Cognito ID code : 

```
- (void)initializeAmazonCognitoProviderWithCompletionBlock:(void (^)(void))completion {
    
    [[UIApplication sharedApplication].keyWindow setUserInteractionEnabled:FALSE];
    NSString* AWSCognitoPoolID = [[[NSBundle mainBundle] infoDictionary] valueForKey:@"AWSCognitoID"];
    AWSCognitoCredentialsProvider *credentialsProvider = [[AWSCognitoCredentialsProvider alloc] initWithRegionType:AMAZON_COGNITO_REGION identityPoolId:AWSCognitoPoolID];
    [credentialsProvider clearCredentials];
    AWSServiceConfiguration *configuration = [[AWSServiceConfiguration alloc] initWithRegion:AWSRegionUSEast1 credentialsProvider:credentialsProvider];
    AWSServiceManager.defaultServiceManager.defaultServiceConfiguration = configuration;
    
    [self getCognitoID:credentialsProvider CompletionBlock:^{
        [self expirationHandler:credentialsProvider CompletionBlock:^{
            dispatch_async(dispatch_get_main_queue(), ^{
              [[UIApplication sharedApplication].keyWindow setUserInteractionEnabled:TRUE];
            });
            completion();
        }];
    }];
    
}

- (void)expirationHandler:(AWSCognitoCredentialsProvider *)creds CompletionBlock: (void (^)(void))completion {
    [[creds credentials] continueWithBlock:^id(AWSTask *task) {
        if (task.error) {
            [self initializeAmazonCognitoProviderWithCompletionBlock:^{}];
        } else {
            AWSCredentials *cred = (AWSCredentials*) task.result;
            NSDateFormatter *dateFormat = [[NSDateFormatter alloc]init];
            [dateFormat setDateFormat:@"yyyy-MM-dd'T'HH:mm:ssZ"];
            /* https://aws.amazon.com/premiumsupport/knowledge-center/security-token-expired/
             https://forums.aws.amazon.com/thread.jspa?threadID=166398
             We should fire timer before 5 minutes of expiration.
             NSString *expF = [dateFormat stringFromDate:cred.expiration];
             */
            [NSTimer scheduledTimerWithTimeInterval:cred.expiration.timeIntervalSinceNow - 300 target:self selector:@selector(initializeAmazonCognitoProviderWithCompletionBlock:) userInfo:nil repeats:NO];
            completion();
        }
        return nil;
    }];
}

- (void)getCognitoID:(AWSCognitoCredentialsProvider *)creds CompletionBlock: (void (^)(void))completion {
    [[creds getIdentityId] continueWithBlock:^id(AWSTask *task) {
        if (task.error) {
            NSLog(@"Error: %@", task.error);
            [self initializeAmazonCognitoProviderWithCompletionBlock:^{}];
        } else {
            NSString *cognitoId = task.result;
            NSUserDefaults *standardDefaults = [NSUserDefaults standardUserDefaults];
            if (![[standardDefaults valueForKey:@"UserCognitoID"] isEqualToString:cognitoId]) {
                [standardDefaults setObject:@"" forKey:BainPreferenceToken];
                [standardDefaults setInteger:0 forKey:@"registrationPostComplete"];
            }
            [standardDefaults setObject:cognitoId forKey:@"UserCognitoID"];
            [standardDefaults synchronize];
            completion();
        }
        return nil;
    }];
}
```

**Comment by rohandubal** [comment]
Hello @dsonara 

To help you better, I would need to understand the use-case you are trying to achieve. 

- Are you using API Gateway with un-authenticated access via Amazon Cognito? 
- Are you expecting any federation/ sign in to occur in the app?
- When you say missing auth token, which token is this?

Also I see that you are clearing credentials every time you initialize. Whats the purpose for that? And why manually handle the refresh?

The SDK has the code to automate fetch of credentials when they are expired to avoid this issue.

Also FYI, we recently released an updated `AWSMobileClient` version `2.7.x` which introduces a lot of new functionality and simplifies API usage. Can you try updating the SDK version to use the updated `AWSMobileClient`?

You can find entire documentation here: https://aws-amplify.github.io/docs/ios/authentication 

Hope this helps.

Thanks,
Rohan

**Comment by dsonara** [comment]
Hello Rohan,

I have been seeking a reply since 2 months. 
I have applied another solution of that. I have been storing credentialProvider object in shared object. And I am using that object for refreshing cognitoID. 

This solution work for me at the moment. I will check for newer version release.

Thank you !!

**Comment by muellerfr** [comment]
Hi @dsonara,

Thanks for letting us know, and great to hear you found a solution. Please try out `AWSMobileClient` as it is the recommended solution. I am going to close this issue. Please open a new issue if you encounter any problems.
